### PR TITLE
Block Collab: New package, a framework for collaborative editing.

### PR DESCRIPTION
**Pull request by epiqueras**
> Make the block editor real-time collaborative.

### CODE REVIEW

**Comment by epiqueras** [comment]
gif

**Comment by epiqueras** [comment]
gif

**Comment by epiqueras** [comment]
cc @dmonad @mtias @shaunandrews @MichaelArestad @pablohoneyhoney

**Comment by MichaelArestad** [comment]
To test, enable the "Collaborative Editing" experiment. It worked like a charm on the first try with multiple windows open.

@epiqueras I have a few questions:
* What options does the package come with?
* Could the cursor be labeled on hover/focus to indicate who is editing?
* Could each user get assigned a unique cursor color?
* Could we show each of the users currently looking at the page? (perhaps as avatars somewhere to be designed)

File: packages/block-collab/src/use-block-collab/index.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	)?.setState;
+	useEffect( () => {
+		// We share instances between hook instances for the same document,
+		// because we need to reuse the same `yDoc` and provider.
+		// Note that the `localYjsChangeSymbol` does differ per hook instance,
+		// so that edits made through one sync to the others.
+		const instanceKey = `${ id }|${ password }`;
+		if ( ! instances[ instanceKey ] ) {
+			const yDoc = new yjs.Doc();
+			const yDocPeers = yDoc.getMap( 'peers' );
```
Comment (dmonad) [inline packages/block-collab/src/use-block-collab/index.js:55]: I suggest using the Awareness protocol instead for showing who is currently working on the document. https://github.com/yjs/y-protocols/

It is a dedicated CRDT for sharing awareness data. The advantage is that it automatically detects when users are offline and that it does not leave any traces in the Yjs document (in the log of changes). You can also have a look at the other editor bindings to see how they make use of the Awareness state. E.g. [y-codemirror](https://github.com/yjs/y-codemirror/blob/master/src/y-codemirror.js)

**Comment by dmonad** [comment]
This is awesome! :heart_eyes: 

This PR seems far more evolved than my PR. Should we close #18357 ?

**Comment by shaunandrews** [comment]
Well... this is impressive. The commenting systems is a great addition, and I like that comments are tied to blocks. But I'm not sure having a new button in every block toolbar is the best way to go about it. Instead, maybe there's a new Comment tool in the Tool menu:

image

Though, I do worry a little about ending up with a Comment highlight indicator on _every_ block. Perhaps a plugin-like sidebar would be better? You could open the sidebar, and then it would list all Comments in the document. If you click one, the block highlights in the canvas, and you see the Comment UI floating near it so you can respond.

--

It would be nice to attach names to the indicators, and maybe introduce colors; Something like this:

image

Or, we could list editors somewhere else in Gutenberg's UI and use shapes along with color to help indicate who's doing what:

image

I think the names might be better in genera, as theres no need to refer to the legend to know who's who.

Perhaps we can fade-out the indicators if someone hasn't made changes after 30+ seconds?

**Comment by MichaelArestad** [comment]
> Perhaps a plugin-like sidebar would be better? You could open the sidebar, and then it would list all Comments in the document. If you click one, the block highlights in the canvas, and you see the Comment UI floating near it so you can respond.

@shaunandrews  +1 The sidebar seems like a logical and familiar place for comments.

> It would be nice to attach names to the indicators

Right. That's why I was asking about the colors and names for indicators. It would be good to show names on indicator hover along with avatars/names/colors somewhere else. I'm not sure about them being in the breadcrumbs toolbar. It's so rare, maybe just showing the first 2-3 avatars in the top bar. If there are more, collapse the rest into a popover?

File: packages/block-collab/src/use-block-collab/index.js (hunk 1 above)
Comment (epiqueras) [reply packages/block-collab/src/use-block-collab/index.js:55]: Ahh, got it. So I would use `setLocalStateField` to set the cursors and `getStates` to read them?

**Comment by enriquesanchez** [comment]
This is really cool 🙌

I agree that having comments on the sidebar would be best. I can see having an additional icon next to 'More tools and options' that opens a comments sidebar.

I also think showing user presence should be more prominent and easier to spot. I thought about having them on top floating slightly above the toolbar, considering that need to accommodate for the top toolbar and narrow widths:

Screen Shot 2020-06-18 at 12 34 21

File: packages/block-collab/src/use-block-collab/index.js (hunk 1 above)
Comment (dmonad) [reply packages/block-collab/src/use-block-collab/index.js:55]: Exactly!

**Comment by mtias** [comment]
> Instead, maybe there's a new Comment tool in the Tool menu:

I agree, the tools menu is underutilized.

@jasmussen also did some designs for collaborative editing a while ago that might still be relevant, including the avatars on the toolbar.

**Comment by jasmussen** [comment]
> @jasmussen also did some designs for collaborative editing a while ago that might still be relevant, including the avatars on the toolbar.

Emphasis "a while ago". It's in #3741 and maybe you best not look 🙈  😅 

The take-way, if anything, is that you can show a list of avatars in the top toolbar, with a small color band at the bottom, and that same color band is used to indicate the block being edited. I would avoid floating the avatars in the canvas, but instead force the top toolbar (if enabled) to stack. I'd also be careful in choosing a palette of beautiful colors to use for this — I or @pablohoneyhoney can help here. Finally, it's worth trying to use a solid line rather than a dashed one, and showing it not-outside the block but in the same way as focus styles are shown at the moment.

**Comment by shaunandrews** [comment]
> is that you can show a list of avatars in the top toolbar

I'm hesitant to put so much focus on avatars when I expect that most WordPress users won't have one. It seems like focusing on showing their Display or User name along with a color — both shown directly on the canvas — would be the best option.

> it's worth trying to use a solid line rather than a dashed one

:100:

> and showing it not-outside the block but in the same way as focus styles are shown at the moment

I'd be fine keeping it consistent with the focus styles, but one concern I have is that with blocks like Image the colored border is very easy to miss.

**Comment by jasmussen** [comment]
> I'd be fine keeping it consistent with the focus styles, but one concern I have is that with blocks like Image the colored border is very easy to miss.

Keeping it flush is mostly about the complexity it avoids, which is the complexity we thankfully managed to reduce with G2. You could presumably tinker with the border width — it's okay that it's a different border width (and yes style if we have to), so as to actually shape-distinguish from focus, which it isn't.

**Comment by chrisvanpatten** [comment]
> I'm hesitant to put so much focus on avatars when I expect that most WordPress users won't have one. It seems like focusing on showing their Display or User name along with a color — both shown directly on the canvas — would be the best option.

This is definitely true, and I don't disagree, but I do want to note there is [an effort to add an avatar upload](https://core.trac.wordpress.org/ticket/16020) (separate from Gravatar) to core which has gained some traction, potentially for 5.5. So it might not necessarily be viable now, but could be in the future.

**Comment by epiqueras** [comment]
Screen Shot 2020-06-22 at 7 03 14 PM

I added colors hashed from the user IDs and started showing the user's name on the margin of the document.

How do we feel about merging this experiment for other people to contribute?

**Comment by jasmussen** [comment]
> How do we feel about merging this experiment for other people to contribute?

If we do get this in, one of the first followups should be a PR to refine the "user is editing" focus style. Merging and iterating is fine when we remember the iteration step — the dashed line and colors used is the only weak point in an otherwise beautiful PR! I have a bit on my plate, otherwise I'd be happy to help.

**Comment by LionsAd** [comment]
Nice work

**Comment by retzion** [comment]
I have not seen anything new on the `add/wordpress-block-collab` branch in over a year. What is the status? Will Gutenberg get collaboration support anytime soon? Thanks!

**Comment by anamwp** [comment]
This is very interesting. 
Anyone working on this `add/wordpress-block-collab` branch. 
I have not seen any progress since Sep 2021. 
Anyone know any update regarding the progress of this branch ? 
Me along with my team are ready to contribute to this PR and we have checked this.
Multi user functionality is not working for us.
Does this PR supported for this ? 
Thanks.

**Comment by sinanisler** [comment]
I tried to build this branch but the build failed. 
anyone can share the build version please ?
thanks

**Comment by sinanisler** [comment]
ok i managed to make a build. insntalled plugin. 

enabled the experimental feature
image

but editor still behaves like normal wp editor. didn't allow multiuser editing same time.

did I miss a step or something?

**Comment by sinanisler** [comment]
looks like there was a dev server for websocket and it is offline now.     :)

**Comment by prajapatisagar** [comment]
Hi team,

I've reviewed the PR, and here are my findings:

Working fine with the old version:
* Real-Time collaboration works on a single browser and with a single user.
* Borders and the user's current type position are functioning as expected.
* Inline comments are working correctly.

However, there are some issues that need to be addressed before we can merge this PR:
* The code in this PR was developed using class components, but we have since moved to function components in the latest Gutenberg block. To make this PR compatible, we need to resolve the conflict and convert the code to support function-based components.
* The default YJS WebSocket ("wss://signaling.yjs.dev/","wss://y-webrtc-signaling-eu.herokuapp.com/","wss://y-webrtc-signaling-us.herokuapp.com/") is no longer working. We should create our own y-js socket and add it here.
* The PR only works in a single browser and is not compatible when we open it in different browsers (Guest or Incognito). We need to ensure the code works across all browsers.
* The current implementation only supports a single user, and it locks posts while another person is editing, displaying the "This post is already being edited" modal. To enhance the user experience, we need to check if real-time collaboration is enabled and avoid displaying the modal in such cases.
* The node version used is 12. We should update the code to be compatible with node versions 14 or 16.

I have a question regarding the implementation using WebSockets. According to the guide (https://make.wordpress.org/core/2023/07/03/real-time-collaboration/), WebRTC doesn't require any extra setup on the server. Do we have any plans on how we're going to handle this?

We can start working on resolving the above points to ensure the PR is ready. Your thoughts and feedback on these changes are welcome.

Thanks

**Comment by youknowriad** [comment]
Hi @prajapatisagar Thanks for the review there. Actually, the exploration have moved now to this PR https://github.com/WordPress/gutenberg/pull/52681 That was a good reminder to close this PR though.

**Pull request closed without merging**